### PR TITLE
OBSDOCS-944 - Power Monitoring - Add Redfish support in power monitoring

### PR DESCRIPTION
**Pull request by shreyasiddhartha**
Change type: Doc update; Power Monitoring - Add Redfish support in power monitoring
Doc JIRA: https://issues.redhat.com/browse/OBSDOCS-944

Fix Version: 4.14+

Doc Preview: https://76231--ocpdocs-pr.netlify.app/openshift-enterprise/latest/observability/power_monitoring/configuring-power-monitoring#power-monitoring-configuring-kepler-redfish_configuring-power-monitoring

SME Review: @sthaha @stleerh @vimalk78 
QE Review: @vprashar2929 
Peer Review: @dfitzmau

### CODE REVIEW

File: modules/power-monitoring-deploying-kepler-redfish.adoc
Comment (sthaha) [inline modules/power-monitoring-deploying-kepler-redfish.adoc:24]: We need to mention that, clicking create will prompt power-monitoring operator to deploy kepler which  in turn creates the `namespace` `openshift-power-montoring` where the secret can be created.

File: modules/power-monitoring-deploying-kepler-redfish.adoc
Comment (sthaha) [inline modules/power-monitoring-deploying-kepler-redfish.adoc:34]: `kepler-operator` isn't the namespace where kepler will be deployed. It is `openshift-power-monitoring`

File: modules/power-monitoring-deploying-kepler-redfish.adoc
Comment (shreyasiddhartha) [reply modules/power-monitoring-deploying-kepler-redfish.adoc:24]: Made some changes to the steps following your comment. PTAL. :)

File: modules/power-monitoring-deploying-kepler-redfish.adoc
Comment (shreyasiddhartha) [reply modules/power-monitoring-deploying-kepler-redfish.adoc]: Changed the namespace to `openshift-power-monitoring`.

File: modules/power-monitoring-deploying-kepler-redfish.adoc
Comment (stleerh) [inline modules/power-monitoring-deploying-kepler-redfish.adoc:24]: Can you specify the secret before it's created?

File: modules/power-monitoring-deploying-kepler-redfish.adoc
Comment (stleerh) [inline modules/power-monitoring-deploying-kepler-redfish.adoc:35]: What should be in redfish.csv?  Should we provide an example?

File: modules/power-monitoring-deploying-kepler-redfish.adoc
Comment (stleerh) [inline modules/power-monitoring-deploying-kepler-redfish.adoc:9]: Under what circumstances?

File: modules/power-monitoring-deploying-kepler-redfish.adoc
Comment (stleerh) [inline modules/power-monitoring-deploying-kepler-redfish.adoc:7]: Since Redfish API isn't always available, how can you check that it's available?

Comment (sthaha) [reply modules/power-monitoring-deploying-kepler-redfish.adoc:7]: This information has to be known to the cluster admin. Kepler does not have a way to detect if redfish is supported on the node.

Comment (sthaha) [inline modules/power-monitoring-deploying-kepler-redfish.adoc:7]:
```suggestion
= Configure {PM-kepler} to use Redfish
```

File: modules/power-monitoring-deploying-kepler-redfish.adoc
Comment (sthaha) [reply modules/power-monitoring-deploying-kepler-redfish.adoc:24]: yes, you can specify the secret that you are going to create (later). Kepler won't be deployed until this secret exists.

File: modules/power-monitoring-deploying-kepler-redfish.adoc
Comment (sthaha) [inline modules/power-monitoring-deploying-kepler-redfish.adoc:31]:
```suggestion
. Create the secret under `openshift-power-monitoring` namespace.
```

File: modules/power-monitoring-deploying-kepler-redfish.adoc
Comment (sthaha) [reply modules/power-monitoring-deploying-kepler-redfish.adoc:35]: +1 We should. Otherwise it is very hard to guess. 

@shreyasiddhartha could you please add the information provide in the jira 

```
redfish.csv Data Format:

your_kubelet_node_name, redfish_username, redfish_password, https://redfish_ip_or_hostname/

Example redfish.csv

kind-control-plane,dummyuser,dummypass,https://dummy-machine.lab.eng.redhat.com
kind-worker,dummyuser,dummypass,https://dummy-machine.lab.eng.redhat.com
kind-worker2,dummyuser,dummypass,https://dummy-machine.lab.eng.redhat.com
```

File: modules/power-monitoring-deploying-kepler-redfish.adoc
Comment (sthaha) [inline modules/power-monitoring-deploying-kepler-redfish.adoc:40]:
```suggestion
Kepler deployment will not progress until the redfish secret is created. This information is reflected in the status of `kepler` instance. 
```

File: modules/power-monitoring-deploying-kepler-redfish.adoc
Comment (sthaha) [reply modules/power-monitoring-deploying-kepler-redfish.adoc:9]:
```suggestion
You can use the Redfish as the source for platform power.
```

**Review by sthaha — CHANGES_REQUESTED** [review]
We need more information about the configuring redfish. Please see the comments. 
Additionally I wonder if the doc should cover the scenario where kepler is already deployed and user wants to configure redfish now.

File: modules/power-monitoring-deploying-kepler-redfish.adoc
Comment (stleerh) [reply modules/power-monitoring-deploying-kepler-redfish.adoc:7]: Is there a command you can run to get this information?

File: modules/power-monitoring-deploying-kepler-redfish.adoc
Comment (sthaha) [inline modules/power-monitoring-deploying-kepler-redfish.adoc:48]:
```suggestion
control-plane,dummyuser,dummypass,https://redfish.nodes.example.com
worker-1,dummyuser,dummypass,https://redfish.nodes.example.com
worker-2,dummyuser,dummypass,https://another.redfish.nodes.example.com
```

File: modules/power-monitoring-deploying-kepler-redfish.adoc
Comment (sthaha) [inline modules/power-monitoring-deploying-kepler-redfish.adoc:53]:
```suggestion
The {PM-kepler} deployment will not continue until the Redfish secret is created. You can find this information in the status of a {PM-kepler} instance.
```

**Review by vprashar2929 — APPROVED** [review]
/lgtm

File: modules/power-monitoring-deploying-kepler-redfish.adoc
Comment (vimalk78) [inline modules/power-monitoring-deploying-kepler-redfish.adoc:35]: we should mention  that 
- the secret must be `Opaque` type
- the secret key must be `redfish.csv`

File: modules/power-monitoring-configuring-kepler-redfish.adoc
Comment (sthaha) [inline modules/power-monitoring-configuring-kepler-redfish.adoc:66]:
```suggestion
$ oc -n openshift-power-monitoring \
      create secret generic redfish-secret \
      --from-file=redfish.csv
```

File: modules/power-monitoring-configuring-kepler-redfish.adoc
Comment (sthaha) [inline modules/power-monitoring-configuring-kepler-redfish.adoc:82]: The rendering seems off. Are we missing a `+` ?

Comment (shreyasiddhartha) [reply modules/power-monitoring-configuring-kepler-redfish.adoc:82]: Yes. fixing this. Thank you!

**Review by sthaha — COMMENTED** [review]
/lgtm

**Review by vprashar2929 — APPROVED** [review]
/lgtm

File: modules/power-monitoring-configuring-kepler-redfish.adoc
Comment (dfitzmau) [inline modules/power-monitoring-configuring-kepler-redfish.adoc:24]:
```suggestion
. Configure `.spec.exporter.redfish` of the {PM-kepler} instance by specifying the mandatory `secretRef` parameter. You can also configure the optional `probeInterval` and `skipSSLVerify` parameters to meet your needs.
```

The customer should decide if they want to configure anything optional. The step seems to indicate that the parameters/fields are needed.

File: modules/power-monitoring-configuring-kepler-redfish.adoc
Comment (dfitzmau) [inline modules/power-monitoring-configuring-kepler-redfish.adoc:39]:
```suggestion
      skipSSLVerify: false <3>
# ...
```

File: modules/power-monitoring-configuring-kepler-redfish.adoc
Comment (dfitzmau) [inline modules/power-monitoring-configuring-kepler-redfish.adoc:35]: I'm not sure what the three dots mean here. The "Example Kepler resource with default configuration" uses a different approach. How about:

deployment:
# ...

File: modules/power-monitoring-configuring-kepler-redfish.adoc
Comment (dfitzmau) [inline modules/power-monitoring-configuring-kepler-redfish.adoc:41]: I think frontloading the callouts with "Required/no required" would help customers instantly know what is mandatory before they read a description.

File: modules/power-monitoring-configuring-kepler-redfish.adoc
Comment (dfitzmau) [inline modules/power-monitoring-configuring-kepler-redfish.adoc:47]:
```suggestion
After the {PM-kepler} is deployed, the  `openshift-power-monitoring` namespace is created.
```

File: modules/power-monitoring-configuring-kepler-redfish.adoc
Comment (dfitzmau) [inline modules/power-monitoring-configuring-kepler-redfish.adoc:49]:
```suggestion
. Create the `redfish.csv` file with the following data format: `<your_kubelet_node_name>, <redfish_username>, <redfish_password>, https://<redfish_ip_or_hostname>/`.
```

File: modules/power-monitoring-configuring-kepler-redfish.adoc
Comment (dfitzmau) [inline modules/power-monitoring-configuring-kepler-redfish.adoc:58]:
```suggestion
. Create the secret under the `openshift-power-monitoring` namespace. You must create the secret with the following conditions:
```

File: modules/power-monitoring-configuring-kepler-redfish.adoc
Comment (dfitzmau) [inline modules/power-monitoring-configuring-kepler-redfish.adoc:59]:
```suggestion
* The secret type is `Opaque`.
```

File: modules/power-monitoring-configuring-kepler-redfish.adoc
Comment (dfitzmau) [inline modules/power-monitoring-configuring-kepler-redfish.adoc:61]: Example heading would be useful.

File: modules/power-monitoring-configuring-kepler-redfish.adoc
Comment (dfitzmau) [inline modules/power-monitoring-configuring-kepler-redfish.adoc:77]:
```suggestion
  redfish.csv: YmFyCg==
  # ...
```

**Comment by dfitzmau** [comment]
Hello @shreyasiddhartha . I added some stylistic suggestions. Nice work on the PR.

File: modules/power-monitoring-configuring-kepler-redfish.adoc
Comment (shreyasiddhartha) [reply modules/power-monitoring-configuring-kepler-redfish.adoc:9]: Kepler is responsible for monitoring the power usage of containers running. I think "power" is the correct usage here.

Comment (dfitzmau) [reply modules/power-monitoring-configuring-kepler-redfish.adoc:9]: Maybe?

You can configure {PM-kepler} to use Redfish as the source for running/hosting containers. {PM-kepler} can then monitor the power usage of these containers

File: modules/power-monitoring-configuring-kepler-redfish.adoc
Comment (shreyasiddhartha) [reply modules/power-monitoring-configuring-kepler-redfish.adoc:24]: I really like your suggestion in here. Is it okay if I use "field" instead of parameters?

File: modules/power-monitoring-configuring-kepler-redfish.adoc
Comment (shreyasiddhartha) [reply modules/power-monitoring-configuring-kepler-redfish.adoc:35]: So, under "deployment", you got 3 more fields which isn't expanded here. The engineering team only wants to show the fields under "redfish". FYI, both are optional tasks and not mandatory for users to perform and Kepler would still be created without specifying "deployment" or "redfish".

File: modules/power-monitoring-configuring-kepler-redfish.adoc
Comment (dfitzmau) [reply modules/power-monitoring-configuring-kepler-redfish.adoc:24]: For sure. Thanks!

File: modules/power-monitoring-configuring-kepler-redfish.adoc
Comment (shreyasiddhartha) [reply modules/power-monitoring-configuring-kepler-redfish.adoc:41]: Rephrased this to:
<1> Required: Specifies the name of the secret that contains the credentials for accessing the Redfish server.
<2> Optional: Controls the frequency at which the power information is queried from Redfish. The default value is `60s`.
<3> Optional: Controls if {PM-kepler} skips verifying the Redfish server certificate. The default value is `false`.

**Comment by shreyasiddhartha** [comment]
> Hello @shreyasiddhartha . I added some stylistic suggestions. Nice work on the PR.

Thanks for the thorough review @dfitzmau. It was really helpful. Made the changes suggested by you. PTAL! :)

File: modules/power-monitoring-configuring-kepler-redfish.adoc
Comment (dfitzmau) [inline modules/power-monitoring-configuring-kepler-redfish.adoc:9]:
```suggestion
You can configure {PM-kepler} to use Redfish as the source for running or hosting containers. {PM-kepler} can then monitor the power usage of these containers.
```

File: modules/power-monitoring-configuring-kepler-redfish.adoc
Comment (dfitzmau) [inline modules/power-monitoring-configuring-kepler-redfish.adoc:49]:
```suggestion
After {PM-kepler} is deployed, the `openshift-power-monitoring` namespace is created.
```

**Comment by dfitzmau** [comment]
Hi @shreyasiddhartha . Thanks for making the updates. I suggested two items that should be fixed before the PR gets merged. Other than these, these LGTM.

File: modules/power-monitoring-configuring-kepler-redfish.adoc
Comment (adellape) [inline modules/power-monitoring-configuring-kepler-redfish.adoc:53]: Same re: not really being "output":
```suggestion
.Example `redfish.csv` file
```

File: modules/power-monitoring-configuring-kepler-redfish.adoc
Comment (adellape) [inline modules/power-monitoring-configuring-kepler-redfish.adoc:58]: There isn't any explicit guidance on the word "dummy" in the ISG or SSG or any of our guides I could find, but could these instances be just "example" instead? Or is there a technical reason it should be "dummy"?
```suggestion
control-plane,exampleuser,examplepass,https://redfish.nodes.example.com
worker-1,exampleuser,examplepass,https://redfish.nodes.example.com
worker-2,exampleuser,examplepass,https://another.redfish.nodes.example.com
```
Just feels similar to the reasoning behind not using "foo"; also a little concerned about potential translation issues / unintended meaning.

File: modules/power-monitoring-configuring-kepler-redfish.adoc
Comment (adellape) [inline modules/power-monitoring-configuring-kepler-redfish.adoc:63]: This unordered list in the middle of the command is making the indentation get off for the command block that follows. You can use an [open block](https://docs.asciidoctor.org/asciidoc/latest/blocks/open-blocks) (`--`) to fix:
```suggestion
. Create the secret under the `openshift-power-monitoring` namespace. You must create the secret with the following conditions:
+
--
* The secret type is `Opaque`.
* The credentials are stored under the `redfish.csv` key in the `data` field of the secret.
--
+
```

File: modules/power-monitoring-configuring-kepler-redfish.adoc
Comment (adellape) [inline modules/power-monitoring-configuring-kepler-redfish.adoc:64]: Don't need a block title for the command, only the example output (per [repo guidelines](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#code-blocks-command-syntax-and-example-output)).
```suggestion
```

File: modules/power-monitoring-configuring-kepler-redfish.adoc
Comment (adellape) [inline modules/power-monitoring-configuring-kepler-redfish.adoc:69]: Just a minor suggestion, lining the spacing up a little differently. 6 spaces seemed like a lot; I believe 4 might be the most common around the repo:
```suggestion
$ oc -n openshift-power-monitoring \
    create secret generic redfish-secret \
    --from-file=redfish.csv
```
or maybe just 2 to line it up with the first command after the prompt:
```suggestion
$ oc -n openshift-power-monitoring \
    create secret generic redfish-secret \
    --from-file=redfish.csv
```

**Review by adellape — COMMENTED** [review]
Some of these are just nits / suggestions, but there are a couple issues that I would like to hold merge on, namely the ["Underscores between multi-word replaceable"](https://github.com/openshift/openshift-docs/pull/76231#discussion_r1621256990) and ["clickable link"](https://github.com/openshift/openshift-docs/pull/76231#discussion_r1621264091) ones.

I'll leave merge-review labels on for now; please let me know when ready for re-review. Thank you!

File: modules/power-monitoring-configuring-kepler-redfish.adoc
Comment (sthaha) [reply modules/power-monitoring-configuring-kepler-redfish.adoc:69]: @adellape , @shreyasiddhartha , my preference would the first version (to **not** lineup) to indicate that line 2 and 3 are continuation of the first line; i.e. its a single command.

**Comment by shreyasiddhartha** [comment]
> Some of these are just nits / suggestions, but there are a couple issues that I would like to hold merge on, namely the ["Underscores between multi-word replaceable"](https://github.com/openshift/openshift-docs/pull/76231#discussion_r1621256990) and ["clickable link"](https://github.com/openshift/openshift-docs/pull/76231#discussion_r1621264091) ones.
> 
> I'll leave merge-review labels on for now; please let me know when ready for re-review. Thank you!

@adellape 

> Some of these are just nits / suggestions, but there are a couple issues that I would like to hold merge on, namely the ["Underscores between multi-word replaceable"](https://github.com/openshift/openshift-docs/pull/76231#discussion_r1621256990) and ["clickable link"](https://github.com/openshift/openshift-docs/pull/76231#discussion_r1621264091) ones.
> 
> I'll leave merge-review labels on for now; please let me know when ready for re-review. Thank you!

@adellape I have made all the other suggestions suggested by you. I'm only leaving the one suggested by @sthaha as it is. Please take a look! Thanks.

File: modules/power-monitoring-configuring-kepler-redfish.adoc
Comment (adellape) [inline modules/power-monitoring-configuring-kepler-redfish.adoc:55]: Sorry, I accidentally left the `\` solution in when I was copy-pasting around in my original comment. The `\` is only necessary when it's in monospace/backticks in-line of a sentence, but not when it's a full codeblock like this:
```suggestion
<your_kubelet_node_name>,<redfish_username>,<redfish_password>,https://<redfish_ip_or_hostname>/
```

File: modules/power-monitoring-configuring-kepler-redfish.adoc
Comment (adellape) [inline modules/power-monitoring-configuring-kepler-redfish.adoc:38]: Still missing the underscore here:
```suggestion
      secretRef: <secret_name> required <1>
```

**Review by adellape — COMMENTED** [review]
Updates are looking good! Need to hold merge for 2 minor nits, however, one of which being my fault.

I will leave the merge-review labels on over the weekend if you'd like for me to merge on Monday, although I am off squad next week (I'm fine with keeping this one, tho). 

Otherwise, if you'd rather re-queue it for someone on-squad next week (who might be in a closer time zone and therefore potentially get back to it quicker than me), feel free to drop the `merge-review-in-progress` label and someone else will see it in the queue and can finish it up. Thanks!

**Comment by shreyasiddhartha** [comment]
> Updates are looking good! Need to hold merge for 2 minor nits, however, one of which being my fault.
> 
> I will leave the merge-review labels on over the weekend if you'd like for me to merge on Monday, although I am off squad next week (I'm fine with keeping this one, tho).
> 
> Otherwise, if you'd rather re-queue it for someone on-squad next week (who might be in a closer time zone and therefore potentially get back to it quicker than me), feel free to drop the `merge-review-in-progress` label and someone else will see it in the queue and can finish it up. Thanks!

@adellape I am okay with you merging this PR, since I have already added in the suggestions. :) Can you please take a look and merge it now? Thank you for all your help.

**Pull request merged**
